services/inc/storage_streams.h

```diff
@@ -0,0 +1,462 @@
+            remaining_ = 0;
+        }
+    }
+
+    virtual ~StorageHalInputStream() = default;
+
+    int read(char* data, size_t size) override {
+        CHECK(peek(data, size));
+        return skip(size);
+    }
+
+    int peek(char* data, size_t size) override {
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        CHECK_TRUE(data, SYSTEM_ERROR_INVALID_ARGUMENT);
+        size = std::min(size, remaining_);
+        CHECK(hal_storage_read(storageId_, address_ + offset_, (uint8_t*)data, size));
+        return size;
+    }
+
+    int skip(size_t size) override {
+        //         size, storageId_, address_, remaining_, offset_);
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        size = std::min(size, remaining_);
+        offset_ += size;
+        remaining_ -= size;
+        return size;
+    }
+
+    int seek(size_t offset) override {
+        //         offset, storageId_, address_, remaining_, offset_);
+        CHECK_TRUE(offset <= (offset_ + remaining_), SYSTEM_ERROR_NOT_ENOUGH_DATA);
+        if (offset < offset_) {
+            remaining_ += (offset_ - offset);
+            offset_ = offset;
+        } else {
+            remaining_ -= (offset - offset_);
+            offset_ = offset;
+        }
+        return offset_;
+    }
+
+    int availForRead() override {
+        return remaining_;
+    }
+
+    int waitEvent(unsigned flags, unsigned timeout) override {
+        if (!flags) {
+            return 0;
+        }
+        if (!(flags & InputStream::READABLE)) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        return InputStream::READABLE;
+    }
+
+private:
+    hal_storage_id storageId_;
+    uintptr_t address_;
+    size_t remaining_;
+    size_t offset_;
+};
+
+#if HAL_PLATFORM_FILESYSTEM
+class FileInputStream : public InputStream {
+public:
+    FileInputStream(const char* filename, filesystem_instance_t instance, size_t offset = 0)
+            : fs_(nullptr),
+              file_{},
+              remaining_(0),
+              offset_(offset),
+              isOpen_(false) {
+        fs_ = filesystem_get_instance(instance, nullptr);
+        open(filename);
+    }
+
+    virtual ~FileInputStream() {
+        if (isOpen_) {
+            isOpen_ = false;
+            fs::FsLock lock(fs_);
+            lfs_file_close(&fs_->instance, &file_);
+        }
+    }
+
+    int read(char* data, size_t size) override {
+        CHECK(peek(data, size));
+        return skip(size);
+    }
+
+    int peek(char* data, size_t size) override {
+        if (!remaining_) {
+            return SYSTEM_ERROR_END_OF_STREAM;
+        }
+        CHECK_TRUE(data, SYSTEM_ERROR_INVALID_ARGUMENT);
+        size = std::min(size, remaining_);
+        fs::FsLock lock(fs_);
+        CHECK(lfs_file_seek(&fs_->instance, &file_, offset_, LFS_SEEK_SET));
```

Comment:
Make sure `CHECK_FS` is used with LittleFS functions here and in other places in this PR.